internal/beater/api/intake/handler.go

```diff
@@ -218,3 +247,61 @@ type jsonError struct {
 	Message  string `json:"message"`
 	Document string `json:"document,omitempty"`
 }
+
+var bodyReaderPool sync.Pool
+
+// pooledBodyReader can be used to consume an io.Reader contents, using a
+// bufio.Reader to avoid reading entire body contents into memory at once.
+// The main use for this reader is to de-couple Request.Body consumption,
+// from the http.Request lifecycle.
+// Once the entire buffer's contents has been consumed (io.EOF), the reader
+// is closed (reset and put back into the pool).
+// pooledBodyReader is not safe for concurrent use.
+type pooledBodyReader struct {
+	bytes.Buffer
+	bufioReader *bufio.Reader
+	closed      bool
+}
+
+func newPooledBodyReader(src io.Reader) (*pooledBodyReader, error) {
+	if b, ok := bodyReaderPool.Get().(*pooledBodyReader); ok {
+		b.reset(src)
+		if _, err := b.Buffer.ReadFrom(b.bufioReader); err != nil {
+			return nil, err
+		}
+		return b, nil
+	}
+	var buf bytes.Buffer
+	bufioReader := bufio.NewReader(src)
+	buf.ReadFrom(bufioReader)
+	return &pooledBodyReader{
+		bufioReader: bufioReader,
+		Buffer:      buf,
+	}, nil
+}
+
+func (b *pooledBodyReader) Read(p []byte) (n int, err error) {
+	n, err = b.Buffer.Read(p)
+	if err != nil && !b.closed {
+		// Close the pooledBodyReader so it can be returned to the pool after
+		// io.EOF is returned (all the contents have been read).
+		if errors.Is(err, io.EOF) {
+			b.Close()
+		}
+	}
```

Comment:
Alternatively, we could not do this in the `Read()` wrapping method and instead wrap the compressed reader with:

```go
// dualCloserReader wraps a two io.ReadCloser using
type dualCloserReader struct {
	// ReadCloser will be used for Reads
	io.ReadCloser
	closer io.Closer
}

//
func newDualCloserReader(reader, closer io.ReadCloser) *dualCloserReader {
	return &dualCloserReader{ReadCloser: reader, closer: closer}
}

// Close closes the two io.ReadCloser
func (d *dualCloserReader) Close() error {
	d.ReadCloser.Close()
	return d.closer.Close()
}
```

Then:
```go
 		reader, err := decoder.CompressedRequestReader(c.Request)
 		if err != nil {
 			writeError(c, compressedRequestReaderError{err})
			return
		}
		// In order to re-use the pooledBodyReaders, they muyst be returned to
		// the pool.
		if c.Async {
			reader = newDualCloserReader(reader, c.Request.Body)
		}
```

Reply:
this isn't applicable anymore.